File: python/create-sql-table.py
import argparse
import csv

def generate_create_table_sql(table_name, columns):
    sql_columns = ", ".join([f"{col} TEXT" for col in columns])
    return f"CREATE TABLE {table_name} (id TEXT PRIMARY KEY, {sql_columns});"

def read_csv_header(file_path):
    with open(file_path, 'r') as csv_file:
        csv_reader = csv.reader(csv_file)
        return next(csv_reader)

def convert_csv_to_sql(input_filename, table_name):
    csv_header = read_csv_header(input_filename)
    property_columns = [col for col in csv_header if col != 'id']

    create_table_sql = generate_create_table_sql(table_name, property_columns)

    print(create_table_sql)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Convert CSV to SQL schema')
    parser.add_argument('--input', required=True, help='Input CSV filename')
    parser.add_argument('--table', required=True, help='Output table name')
    args = parser.parse_args()

    input_filename = args.input
    table_name = args.table

    convert_csv_to_sql(input_filename, table_name)

File: node/create-sql-table.js
const fs = require('fs');
const path = require('path');
const argparse = require('argparse');
const csv = require('csv-parser');

function generateCreateTableSql(tableName, columns) {
  const sqlColumns = columns.map((col) => `${col} TEXT`).join(', ');
  return `CREATE TABLE PUBLIC.${tableName} (id TEXT PRIMARY KEY, ${sqlColumns});`;
}

function readCsvHeader(filePath) {
  return new Promise((resolve, reject) => {
    const stream = fs.createReadStream(filePath);
    stream.pipe(csv())
      .on('headers', (headers) => {
        const cleanedHeaders = headers.map(header => 
          header.replace(/ /g, '_').replace(/\(|\)/g, '')
        );
        stream.destroy();
        resolve(cleanedHeaders);
      })
      .on('error', (err) => {
        reject(err);
      });
  });
}

async function convertCsvToSql(inputFilename, tableName) {
  const csvHeader = await readCsvHeader(inputFilename);
  const propertyColumns = csvHeader.filter((col) => col !== 'id');

  const createTableSql = generateCreateTableSql(tableName, propertyColumns);

  console.log(createTableSql);
}

const parser = new argparse.ArgumentParser({
  description: 'Convert CSV to SQL schema',
});
parser.addArgument(['--input'], { required: true, help: 'Input CSV filename' });
parser.addArgument(['--table'], { required: true, help: 'Output table name' });
const args = parser.parseArgs();

const inputFilename = args.input;
const tableName = args.table;

convertCsvToSql(inputFilename, tableName);
